include/seqan3/io/sam_file/output.hpp

```diff
@@ -150,7 +151,27 @@ class sam_file_output
     //!\brief Move assignment is defaulted.
     sam_file_output & operator=(sam_file_output &&) = default;
     //!\brief Destructor is defaulted.
-    ~sam_file_output() = default;
+    ~sam_file_output()
+    {
+        if (no_record_has_been_written)
+        {
+            assert(!format.valueless_by_exception());
+
+            std::visit(
+                [&](auto & f)
+                {
+                    if constexpr (std::same_as<ref_ids_type, ref_info_not_given>)
+                    {
+                        f.write_header(*secondary_stream, options, std::ignore);
+                    }
+                    else
+                    {
+                        f.write_header(*secondary_stream, options, *header_ptr);
+                    }
+                },
+                format);
+        }
+    };
```

Comment:
```suggestion
    ~sam_file_output()
    {
        if (!no_record_has_been_written)
            return;

        assert(!format.valueless_by_exception());

        std::visit(
            [&](auto & f)
            {
                if constexpr (std::same_as<ref_ids_type, ref_info_not_given>)
                    f.write_header(*secondary_stream, options, std::ignore);
                else
                    f.write_header(*secondary_stream, options, *header_ptr);
            },
            format);
    }
```

Maybe? Anyway, the `};` in the last line is wrong because it actually is just the end of the deconstructor body.

If we do it like this, we might flip the meaning of `no_record_has_been_written`, i.e. change it to `record_has_been_written`.
Instead of `record`, we maybe want to write `header`? Because that it what we actually use it to check for.